adaptive/learner/balancing_learner.py

```diff
@@ -111,38 +111,57 @@ def strategy(self, strategy):
             raise ValueError(
                 'Only strategy="loss_improvements", strategy="loss", or'
                 ' strategy="npoints" is implemented.')
+        self._points = {}  # reset the cache
 
     def _ask_and_tell_based_on_loss_improvements(self, n):
-        points = []
-        loss_improvements = []
+        chosen_points = []
+        chosen_loss_improvements = []
+        npoints_per_learner = defaultdict(int)
+
         for _ in range(n):
             improvements_per_learner = []
-            pairs = []
+            points_per_learner = []
             for index, learner in enumerate(self.learners):
                 if index not in self._points:
                     self._points[index] = learner.ask(
                         n=1, tell_pending=False)
-                point, loss_improvement = self._points[index]
-                improvements_per_learner.append(loss_improvement[0])
-                pairs.append((index, point[0]))
-            x, l = max(zip(pairs, improvements_per_learner),
-                       key=itemgetter(1))
-            points.append(x)
-            loss_improvements.append(l)
-            self.tell_pending(x)
-
-        return points, loss_improvements
+                points, loss_improvements = self._points[index]
```

Comment:
This is really confusing and seems like we've chosen a bad name for `self._points`: from that name it's not at all clear that loss improvements are stored, neither it is clear what these `self._points` actually are. All points ever requested?

Reply:
It's an cache for `ask`, so maybe `_ask_cache`?

Reply:
Yes, that's much better.